# Green Revenue CFS API Workflow (CSV)

Example Code Disclaimer: ALL EXAMPLE CODE IS PROVIDED ON AN “AS IS” AND “AS AVAILABLE” BASIS FOR ILLUSTRATIVE PURPOSES ONLY. LSEG MAKES NO REPRESENTATIONS OR WARRANTIES OF ANY KIND, EXPRESS OR IMPLIED, AS TO THE OPERATION OF THE EXAMPLE CODE, OR THE INFORMATION, CONTENT, OR MATERIALS USED IN CONNECTION WITH THE EXAMPLE CODE. YOU EXPRESSLY AGREE THAT YOUR USE OF THE EXAMPLE CODE IS AT YOUR SOLE RISK.

### Importing libararies

In [1]:
import os
import sys
import requests 
import json
from dotenv import dotenv_values
config = dotenv_values(".env")

### Set RDP credentials and Initial Parameters

In [2]:
username = config['RDP_USERNAME'] #or replacing with your RDP Username/Machine-ID
password = config['RDP_PASSWORD'] #or replacing with your RDP Password
clientId = config['RDP_APP_KEY'] #or replacing with your RDP App key

RDP_HOST= 'https://api.refinitiv.com'
acccess_token = None
refresh_token = None
expires_in = 0

## <a id="rdp_workflow"></a>RDP APIs Application Workflow

### Step 1: Authentication with RDP APIs

Refinitiv Data Platform entitlement check is based on OAuth 2.0 specification. The first step of an application workflow is to get a token from RDP Auth Service, which will allow access to the protected resource, i.e. data REST API. 

The API requires the following access credential information:
- Username: The username. 
- Password: Password associated with the username. 
- Client ID: This is also known as ```AppKey```, and it is generated using an App key Generator. This unique identifier is defined for the user or application and is deemed confidential (not shared between users). The client_id parameter can be passed in the request body or as an “Authorization” request header that is encoded as base64.

The HTTP request for the RDP APIs Authentication service is as follows:

``` HTTP
POST /auth/oauth2/v1/token HTTP/1.1
Accept: */*
Content-Type: application/x-www-form-urlencoded
Host: api.refinitiv.com:443
Content-Length: XXX

username=RDP_USERNAME
&password=RDP_PASSWORD
&client_id=RDP_APP_KEY
&grant_type=password
&takeExclusiveSignOnControl=true
&scope=trapi
```

Once the authentication success, the function gets the RDP Auth service response message and keeps the following RDP token information in the variables.
- **access_token**: The token used to invoke REST data API calls as described above. The application must keep this credential for further RDP APIs requests.
- **refresh_token**: Refresh token to be used for obtaining an updated access token before expiration. The application must keep this credential for access token renewal.
- **expires_in**: Access token validity time in seconds.

Next, after the application received the Access Token (and authorization token) from RDP Auth Service, all subsequent REST API calls will use this token to get the data. Please find more detail regarding RDP APIs workflow in the following resources:
- [RDP APIs: Introduction to the Request-Response API](https://developers.refinitiv.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/tutorials#introduction-to-the-request-response-api) page.
- [RDP APIs: Authorization - All about tokens](https://developers.refinitiv.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/tutorials#authorization-all-about-tokens) page.

In [3]:
#step 1 - get RDP Access Token from RDP

import requests 
import json

# Send HTTP Request
auth_url = f'{RDP_HOST}/auth/oauth2/v1/token'
payload = f'grant_type=password&username={username}&client_id={clientId}&password={password}&takeExclusiveSignOnControl=True&scope=trapi'
try:
    response = requests.post(auth_url, 
                             headers = {'Content-Type':'application/x-www-form-urlencoded'}, 
                             data = payload, 
                             auth = (clientId, '')
                )
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')


if response.status_code == 200:  # HTTP Status 'OK'
    print('Authentication success')
    access_token = response.json()['access_token']
    refresh_token = response.json()['refresh_token']
    expires_in = int(response.json()['expires_in'])

if response.status_code != 200:
    print(f'RDP authentication failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Authentication success


## <a id="rdp_get_data"></a>Requesting Data from RDP APIs

That brings us to requesting the RDP APIs data. All subsequent REST API calls use the Access Token via the *Authorization* HTTP request message header as shown below to get the data. 
- Header: 
    * Authorization = ```Bearer <RDP Access Token>```

Please notice *the space* between the ```Bearer``` and ```RDP Access Token``` values.

The application then creates a request message in a JSON message format or URL query parameter based on the interested service and sends it as an HTTP request message to the Service Endpoint. Developers can get RDP APIs the Service Endpoint, HTTP operations, and parameters from Refinitiv Data Platform's [API Playground page](https://api.refinitiv.com/) - which is an interactive documentation site developers can access once they have a valid Refinitiv Data Platform account.

## <a id="rdp_cfs_data"></a>Requesting CFS Data

### Step 2: Listing the Package Ids using the Bucket Name

**Note**: **If you already know your package Ids, you can skip to #step3**

To request the CFS data, the first step is to send an HTTP ```GET``` request to the RDP ```/file-store/v1/packages?bucketName={bucket-name}``` endpoint to list all Package Ids under the input ```bucket-name```.

The HTTP Request structure is as follows:

``` HTTP
GET /file-store/v1/packages?bucketName={bucket-name} HTTP/1.1
Host: api.refinitiv.com
Authorization: Bearer <Access Token>
```
**Note**: The bucket name is *case-insensitive*.

The Green Revenue bucket name is **bulk-greenrevenue**.

In [4]:
#step 2 - list Package IDs from bucket name

CFS_url = f'{RDP_HOST}/file-store/v1/packages?bucketName=bulk-greenrevenue'

try:
    response = requests.get(CFS_url, headers={'Authorization': f'Bearer {access_token}'})
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')

if response.status_code == 200:  # HTTP Status 'OK'
    print('Receive list Package IDs from RDP APIs')
else:
    print(f'RDP APIs: CFS request failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Receive list Package IDs from RDP APIs


Example of the first entry of package IDs, the pacakgeId is the ```packageId``` field.

In [5]:
print(json.dumps(response.json()['value'][0], sort_keys=True, indent=2, separators=(',', ':')))

{
  "bucketNames":[
    "bulk-GreenRevenue"
  ],
  "contactEmail":"naveen@lseg.com",
  "created":"2022-09-30T13:45:54Z",
  "modified":"2023-02-10T09:41:50Z",
  "packageId":"4316-d43b-81c40763-8e6a-0dbec8162ab1",
  "packageName":"Bulk-GR-Global-Standard-Full-v1",
  "packageType":"bulk"
}


The next step is choosing the package Id. Based on the [Green Revenues User Guide](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/documentation#green-revenues-user-guide), the package Ids are the following:
- Package Name: Bulk-GR-Global-Summary-Full-v1: package Id *4e94-6d63-fea034dc-90e2-de33895bd4e9*
- Package Name: Bulk-GR-Global-Standard-Full-v1: package Id *4316-d43b-81c40763-8e6a-0dbec8162ab1*

In [6]:
packageId = response.json()['value'][0]['packageId']
packageId

'4316-d43b-81c40763-8e6a-0dbec8162ab1'

### Step 3: Listing the Filesets of the Bulk Green Revenues Data with the packageId

The next step is calling the CFS API with the buket name and package Id to list all FileSets using ```bucket-name``` and ```packageId```.

API endpint is ```/file-store/v1/file-sets?bucket=bulk-greenrevenue&packageId={packageId}```

The HTTP Request structure is as follows:

``` HTTP
GET /file-store/v1/file-sets?bucket=bulk-greenrevenue&packageId={packageId} HTTP/1.1
Host: api.refinitiv.com
Authorization: Bearer <Access Token>
```

In [7]:
#step 3 - get file id from bucket name

CFS_url = f'{RDP_HOST}/file-store/v1/file-sets?bucket=bulk-greenrevenue&packageId={packageId}'

try:
    response = requests.get(CFS_url, headers={'Authorization': f'Bearer {access_token}'})
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')


if response.status_code == 200:  # HTTP Status 'OK'
    print('Receive FileSets list from RDP APIs')
else:
    print(f'RDP APIs: CFS request failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Receive FileSets list from RDP APIs


In [8]:
print(json.dumps(response.json()['value'][0], sort_keys=True, indent=2, separators=(',', ':')))

{
  "attributes":[
    {
      "name":"ContentType",
      "value":"GR Global Standard Full"
    }
  ],
  "availableFrom":"2023-11-19T21:06:01Z",
  "availableTo":"2023-12-19T21:06:01Z",
  "bucketName":"bulk-GreenRevenue",
  "contentFrom":"1970-01-01T00:00:00Z",
  "contentTo":"2023-11-19T20:55:00Z",
  "created":"2023-11-19T21:06:01Z",
  "files":[
    "45df-a5cd-ce3ed317-a215-68a38bc24876"
  ],
  "id":"40f4-0969-f632f924-bcee-654ebcb1f8f2",
  "modified":"2023-11-19T21:06:12Z",
  "name":"Bulk-GR-Global-Standard-Full-v1-Jsonl-Init-2023-11-19T21:03:20.924Z",
  "numFiles":1,
  "packageId":"4316-d43b-81c40763-8e6a-0dbec8162ab1",
  "status":"READY"
}


The File ID is in the ```files``` array.

Next step is fine the files in CSV format from the return JSON 'name' attribute. This demonstration just uses the first one that has been found.

In [9]:
index = 0
for index, package in enumerate(response.json()['value']):
    if 'Csv' in package['name']:
        print(index, package['name'])
        break

2 Bulk-GR-Global-Standard-Full-v1-Csv-Init-2023-12-03T21:01:53.989Z


In [10]:
file_id = response.json()['value'][index]['files'][0]
file_id

'4977-f997-60aa082b-b889-3b5bcd558c61'

# Step 3.1: Paging

By default, the ```/file-store/v1/file-sets``` and ```/file-store/v1/packages?bucketName={bucket-name}``` endpoints always return 25 results per request. You can adjust the number of return results via the ```pageSize``` query parameter, the maximum number is **100**.

The result also has the ```@nextLink``` node that contains the URL for requesting the next page of query.

You can find more detail about the Paging and @nextLink node on the step 3.1 of the [A Step-By-Step Workflow Guide for RDP Client File Store (CFS) API](https://developers.lseg.com/en/article-catalog/article/a-step-by-step-workflow-guide-for-rdp-client-file-store--cfs--ap) article and [GitHub](https://github.com/LSEG-API-Samples/LSEG-API-Samples-Example.RDP.Python.GenericBulkFile.Workflow).


### Step 4: Get the file URL on AWS S3

The last step is downloading the FIle using File ID with the RDP ```/file-store/v1/files/{file ID}/stream``` endpoint.

The HTTP Request structure is as follows:

``` HTTP
GET /file-store/v1/files/{fileId}/stream?doNotRedirect=true HTTP/1.1
Host: api.refinitiv.com
Authorization: Bearer <Access Token>
```

In [11]:
#step 4 - get file URL from file id

FileID_url = f'{RDP_HOST}/file-store/v1/files/{file_id}/stream?doNotRedirect=true'

try:
    response = requests.get(FileID_url, headers={'Authorization': f'Bearer {access_token}'})
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')


if response.status_code == 200:  # HTTP Status 'OK'
    print('Receive File URL from RDP APIs')
else:
    print(f'RDP APIs: CFS request failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Receive File URL from RDP APIs


The File URL is in the ```url``` attribute.

In [12]:
file_url = response.json()['url']
file_url

'https://a206464-bulk-greenrevenue.s3.amazonaws.com/Bulk-GR-Global-Standard-Full-v1/2023/12/03/Bulk-GR-Global-Standard-Full-v1-Init-2023-12-03T21%3A01%3A53.989Z.csv.gz?x-request-Id=bcb64cd2-99c5-4f5b-ab6d-a70012ba0e1b&x-package-id=4316-d43b-81c40763-8e6a-0dbec8162ab1&x-client-app-id=b4842f3904fb4a1fa18234796368799086c63541&x-file-name=Bulk-GR-Global-Standard-Full-v1-Init-2023-12-03T21%3A01%3A53.989Z.csv.gz&x-fileset-id=4167-9a5b-8a7c8b6c-a2fe-e3c1006b324b&x-bucket-name=bulk-GreenRevenue&x-uuid=GESG1-103676&x-file-Id=4977-f997-60aa082b-b889-3b5bcd558c61&x-fileset-name=Bulk-GR-Global-Standard-Full-v1-Csv-Init-2023-12-03T21%3A01%3A53.989Z&x-event-external-name=cfs-claimCheck-download&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEPf%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQDVLdH0AJkdFycgAuHGlpZurkJRBzrhACTyPh%2F1Q%2BPcVwIhAKr%2FtdwGM46D5%2FWBK2JEB6i%2F6X%2BwPB6WjQttJiMfeRRnKpoCCHAQBBoMNjQyMTU3MTgxMzI2Igzo99ligNebakbwZEMq9wGJ7PJqaKFUFE%2FFl4tbGDslgfXtEuZb6CvHXlOEjCRJutIPdriVJ216DMPLLeU

### Step 5: Downloading the file

Based on the S3 ```file_url``` above, the actual file name is *Bulk-GR-Global-Summary-Full-v1-Init-2023-10-01T21_01_18.831Z.csv.gz*. So you need to replace the escape character ```%3A``` with ```_``` (underscore) character.

**Note**: If you cannot download the file, please wait for a while and then retry download the file from the URL. Please do not flush the download requests.

In [14]:
#step 5 - Downlaod file
import polling2

zipfilename = file_url.split("?")[0].split("/")[-1].replace("%3A","_")
print(f'Downloading File {zipfilename} ...')

def test_result(response):
    return response.status_code == 200

try:
    response = polling2.poll(lambda: requests.get(file_url), 
                            step = 10,
                            poll_forever = True,
                            check_success= test_result)
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')

if response.status_code == 200:  # HTTP Status 'OK'
    print('Receive File Successfully')
    open(zipfilename, 'wb').write(response.content)
    print(f'{zipfilename} Saved')
else:
    print(f'RDP APIs: Request file failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Receive File Successfully
Bulk-GR-Global-Standard-Full-v1-Init-2023-12-03T21_01_53.989Z.csv.gz Saved


And then unzip the file.

In [15]:
#unzip file
import gzip
import shutil
try:
    unzipfilename = zipfilename.split('.gz')[0]
    print(f'Unzip to {unzipfilename} ...')
    with gzip.open(zipfilename, 'rb') as f_in:
        with open(unzipfilename, 'wb') as f_out:
            shutil.copyfileobj(f_in, f_out)
    print('Done')
except Exception as e:
    print('The error is: ',e)

Unzip to Bulk-GR-Global-Standard-Full-v1-Init-2023-12-03T21_01_53.989Z.csv ...
Done


In [16]:
# View some data
with open(unzipfilename) as f:
    data = f.read(2000)
    print(data)

_Row_Type,ObjectId,GreenRevenue_SectorRevenues_ParentSectorCode,GreenRevenue_SectorRevenues_ParentSectorCodeDesription,GreenRevenue_SectorRevenues_SectorCode,GreenRevenue_SectorRevenues_SectorCodeDescription,GreenRevenue_SectorRevenues_SectorRevenuePercent,GreenRevenue_SectorRevenues_SectorTypeCode
"",4295877937;293,TE.03,Transport Equipment Road Vehicles,TE.03.4,Transport Equipment Electrified Road Vehicles & Devices (inc Hydrogen powered),14.5236,Microsector
"",4295877937;293,WI.05,Water Infrastructure & Technologies Natural Disaster Response,WI.05.0,Water Infrastructure & Technologies Natural Disaster Response (General),0,Microsector
"",4295877937;293,null,null,TE,Transport Equipment,14.5236,Sector
"",4295877937;293,null,null,FA,Food & Agriculture,0,Sector
"",4295877937;293,null,null,EM,Energy Management & Efficiency,4.7389,Sector
"",4295877937;293,null,null,WI,Water Infrastructure & Technologies,0,Sector
"",4295877937;293,FA.04,Food & Agriculture Logistics,FA.04.0,Food & Agricultur

### Step 6: Refresh Token with RDP APIs

Before the session expires (based on the ```expires_in``` parameter, in seconds) , an application needs to send a Refresh Grant request message to RDP Authentication service to get a new access token before further request data from the platform.

The API requires the following access credential information:
- Refresh Token: The current Refresh Token value from the previous RDP Authentication call
- Client ID: This is also known as ```AppKey```, and it is generated using an App key Generator. This unique identifier is defined for the user or application and is deemed confidential (not shared between users). The client_id parameter can be passed in the request body or as an “Authorization” request header that is encoded as base64.
- Grant Type ```refresh_token```: This is for getting a new Access Token. 

The HTTP request for the RDP APIs Authentication service is as follows:

``` HTTP
POST /auth/oauth2/v1/token HTTP/1.1
Accept: */*
Content-Type: application/x-www-form-urlencoded
Host: api.refinitiv.com:443
Content-Length: XXX

refresh_token={current_refresh_token}
&grant_type=refresh_token
&client_id=RDP_APP_KEY
```

Once the authentication success, the function gets **access_token**, **refresh_token**, and **expires_in** from the RDP Auth service response message the same as the previous RDP Authentication call. An application must keep those value for the next Refresh Token call.

#### Caution: API Limit

The RDP Authentication service has the API limit described on the [RDP APIs: Limitations and Guidelines for the RDP Authentication Service](https://developers.lseg.com/en/article-catalog/article/limitations-and-guidelines-for-the-rdp-authentication-service) article.  If the application flushes the authentication request messages (both ```password``` and ```refresh_token``` grant_type) beyond the limit, the account will be blocked by the API Gateway. 

In [17]:
#step 6 - Refreshing Token

# Send HTTP Request
auth_url = f'{RDP_HOST}/auth/oauth2/v1/token'
payload = f'grant_type=refresh_token&client_id={clientId}&refresh_token={refresh_token}'
try:
    response = requests.post(auth_url, 
                             headers = {'Content-Type':'application/x-www-form-urlencoded'}, 
                             data = payload, 
                             auth = (clientId, '')
                )
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')

if response.status_code == 200:  # HTTP Status 'OK'
    print('Refresh Token success')
    access_token = response.json()['access_token']
    refresh_token = response.json()['refresh_token']
    expires_in = int(response.json()['expires_in'])

if response.status_code != 200:
    print(f'RDP authentication failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Refresh Token success


### Step 7: Revoke Token to ending the session.

This revocation mechanism allows an application to invalidate its tokens if the end-user logs out, changes identity, or exits the respective application. Notifying the authorization server that the token is no longer needed allows the authorization server to clean up data associated with that token (e.g., session data) and the underlying authorization grant.

The API requires the following HTTP Header and Credential parameter information:
- Header: 
    * Authorization = ```Basic <App Key in Base64 format>```

Please notice *the space* between the ```Basic``` and ```App Key in Base64 format``` values.
- Body parameter
    * token: The current ```Access Token``` value from the previous RDP Authentication call

The HTTP request for the RDP APIs Authentication service is as follows:

``` HTTP
POST /auth/oauth2/v1/revoke HTTP/1.1
Accept: */*
Content-Type: application/x-www-form-urlencoded
Host: api.refinitiv.com:443
Authorization: Basic <App Key in Base64>
Content-Length: XXX

token={current_Access_token}
```

In [18]:
#step 7 - Revoking Token

import base64

clientId_bytes = clientId.encode('ascii')
base64_bytes = base64.b64encode(clientId_bytes)
clientId_base64 = base64_bytes.decode('ascii')

# Send HTTP Request
auth_url = f'{RDP_HOST}/auth/oauth2/v1/revoke'
payload = f'token={access_token}'
try:
    response = requests.post(auth_url, 
                             headers = {
                                 'Content-Type':'application/x-www-form-urlencoded',
                                 'Authorization': f'Basic {clientId_base64}'
                             }, 
                             data = payload, 
                             auth = (clientId, '')
                )
except requests.exceptions.RequestException as exp:
    print(f'Caught exception: {exp}')

if response.status_code == 200:  # HTTP Status 'OK'
    print('Revoke Token success')
if response.status_code != 200:
    print(f'RDP authentication failure: {response.status_code} {response.reason}')
    print(f'Text: {response.text}')

Revoke Token success


## <a id="references"></a>References

That brings me to the end of my Green Revenue CFS workflow project. For further details, please check out the following resources:
* [Refinitiv Data Platform APIs page](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis) on the [Refinitiv Developer Community](https://developers.refinitiv.com/) website.
* [Refinitiv Data Platform APIs Playground page](https://api.refinitiv.com).
* [Refinitiv Data Platform APIs: Introduction to the Request-Response API](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/tutorials#introduction-to-the-request-response-api).
* [Refinitiv Data Platform APIs: Authorization - All about tokens](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/tutorials#authorization-all-about-tokens).
* [Limitations and Guidelines for the RDP Authentication Service](https://developers.lseg.com/en/article-catalog/article/limitations-and-guidelines-for-the-rdp-authentication-service) article.
* [Getting Started with Refinitiv Data Platform](https://developers.lseg.com/en/article-catalog/article/getting-start-with-refinitiv-data-platform) article.
* [RDP CFS service (general) user guide](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/documentation#cfs-api-user-guide)
* [RDP Green Revenue user guide](https://developers.lseg.com/en/api-catalog/refinitiv-data-platform/refinitiv-data-platform-apis/documentation#green-revenues-user-guide)

For any questions related to Refinitiv Data Platform APIs, please use the [RDP APIs Forum](https://community.developers.refinitiv.com/spaces/231/index.html) on the [Developers Community Q&A page](https://community.developers.refinitiv.com/).